### PR TITLE
Fixes tesla bugs, makes code better

### LINKED ISSUE #45908: ethereals are apparently tesla proof
## Reproduction: be ethereal 2. spawn tesla ball 3. be unharmed???

not 100% sure on this but wow that is way too overpowered if true, last round was a single ethereal with space adaptation surviving a tesla directly on top of the shuttle

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Tesla fixes galore, fixes #45908, the main zap was using TRUE to detremine it's flags. so the flags looked like 0x00001. This meant the tesla shock flag was not set, and so no damage.
WHY, WHO, HOW? I HAVE NO CLUE.

Fixes some runtimes.
Removes some of my shitcode and makes zap_act fully separate.
Makes zap splitting less common, and makes range dependent on the thing you hit.
#23676 broke #23678, this fixes the effect.
Makes explosions stop a zap chain.

## Why It's Good For The Game

Fixes man good.

## Changelog
:cl:
add: Tesla zaps will now turn tesla generators into teslas.
fix: Explosions might be a little less common, I've fixed the thing where machines would explode with no visible arc to the object. 
fix: Fixes a whole bunch of issues, teslas make more sense now
code: Makes my shitcode better
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Will make a pr to reduce tesla lag in future, will most likely remove explosions.

**Comment by trollbreeder** [comment]
> Will make a pr to reduce tesla lag in future, will most likely remove explosions.

No, not my tesloose explosions!

File: code/modules/power/tesla/energy_ball.dm [hunk 1]
```diff
@@ -169,6 +169,8 @@
 	C.dust()
 
 /proc/tesla_zap(atom/source, zap_range = 3, power, zap_flags = ZAP_DEFAULT_FLAGS, list/shocked_targets)
+	if(source == null)
```
Comment (yoyobatty) [inline code/modules/power/tesla/energy_ball.dm]: You could do QDELETED instead since it checks if it exists or if it's a null

Comment (LemonInTheDark) [reply code/modules/power/tesla/energy_ball.dm]: Sure, why not

**Comment by PKPenguin321** [comment]
dont remove explosions

**Comment by LemonInTheDark** [comment]
> 
> 
> dont remove explosions

I'll try

**Comment by LemonInTheDark** [comment]
Shit fuck and also damnit.
Previously undocced change, before this fix explosions would happened seemingly randomly, this is because a zap would hit a machine, jump. and then the machine would blow up, which meant it could not be used to draw off of. This is fixed, but this also means that each zap can cause at max one explosion, after which it terminates.

File: code/game/objects/obj_defense.dm [hunk 2]
```diff
@@ -229,12 +229,10 @@ GLOBAL_DATUM_INIT(acid_overlay, /mutable_appearance, mutable_appearance('icons/e
 ///Called when the obj is hit by a tesla bolt.
 /obj/proc/zap_act(power, zap_flags, shocked_targets)
 	if(QDELETED(src))
-		return
+		return 0
```
Comment (spookydonut) [inline code/game/objects/obj_defense.dm]: nothing checks return value

File: code/modules/power/tesla/energy_ball.dm [hunk 3]
```diff
@@ -312,36 +313,46 @@
...
 
 	else if(!QDELETED(closest_structure))
-		closest_structure.zap_act(power, zap_flags, shocked_targets)
+		power = closest_structure.zap_act(power, zap_flags, shocked_targets)
+
+
+	var/zap_count = 1
+	if(prob(20))
+		zap_count++
+	for(var/i in 1 to zap_count)
```
Comment (spookydonut) [inline code/modules/power/tesla/energy_ball.dm]:
```suggestion
	for(var/i in 1 to pick(1, 2; 20))
```

File: code/game/objects/obj_defense.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/game/objects/obj_defense.dm]: All of the power assignments check the return value, so I put the 0 there as I'm not sure what the default return type is, and wanted to be sure.

Comment (spookydonut) [reply code/game/objects/obj_defense.dm]: nothing that uses zap_act() that i can see checks the return value of it, even parent calls of zap_act() dont check parent return value

File: code/modules/power/tesla/energy_ball.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/modules/power/tesla/energy_ball.dm]: What wizardry is this.

File: code/game/objects/obj_defense.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/game/objects/obj_defense.dm]: Unless I am doing this incorrectly these should be using the return value.
image

Comment (spookydonut) [reply code/game/objects/obj_defense.dm]: oh i missed those.

File: code/modules/power/tesla/energy_ball.dm (hunk 3 above)
Comment (AnturK) [reply code/modules/power/tesla/energy_ball.dm]: Uh, i'm pretty sure the format is other way around ? As in this is prob(2) of 20 ?

Comment (spookydonut) [reply code/modules/power/tesla/energy_ball.dm]: oh oops yeah

**Pull request merged**